File: backend/app/api/v1/categories.py
"""Category endpoints."""

from uuid import UUID

from fastapi import APIRouter

from app.api.deps import CurrentUser, DbSession, OrgContext
from app.schemas.domain import CategoryCreate, CategoryResponse, CategoryUpdate
from app.services import category_service

router = APIRouter(prefix="/categories", tags=["categories"])


@router.get("", response_model=list[CategoryResponse])
async def list_categories(
    db: DbSession,
    user: CurrentUser,
    org_ctx: OrgContext,
) -> list[CategoryResponse]:
    org, _ = org_ctx
    categories = await category_service.get_categories(db, org.id)
    return [CategoryResponse.model_validate(c) for c in categories]


@router.post("", response_model=CategoryResponse, status_code=201)
async def create_category(
    body: CategoryCreate,
    db: DbSession,
    user: CurrentUser,
    org_ctx: OrgContext,
) -> CategoryResponse:
    org, _ = org_ctx
    cat = await category_service.create_category(
        db,
        org_id=org.id,
        name=body.name,
        kind=body.kind,
        color=body.color,
        icon=body.icon,
    )
    await db.commit()
    await db.refresh(cat)
    return CategoryResponse.model_validate(cat)


@router.patch("/{category_id}", response_model=CategoryResponse)
async def update_category(
    category_id: UUID,
    body: CategoryUpdate,
    db: DbSession,
    user: CurrentUser,
    org_ctx: OrgContext,
) -> CategoryResponse:
    org, _ = org_ctx
    cat = await category_service.update_category(
        db,
        org_id=org.id,
        category_id=category_id,
        patch=body.model_dump(exclude_unset=True),
    )
    await db.commit()
    await db.refresh(cat)
    return CategoryResponse.model_validate(cat)


@router.delete("/{category_id}", status_code=204)
async def delete_category(
    category_id: UUID,
    db: DbSession,
    user: CurrentUser,
    org_ctx: OrgContext,
) -> None:
    org, _ = org_ctx
    await category_service.delete_category(
        db,
        org_id=org.id,
        category_id=category_id,
    )
    await db.commit()

File: frontend/src/lib/api/categories.ts
import { apiFetch } from "@/lib/api/client";

export type CategoryKind = "income" | "expense";

export type Category = {
  id: string;
  organization_id: string;
  name: string;
  kind: CategoryKind;
  is_default: boolean;
  created_at: string;
  updated_at: string;
};

export type CategoryCreate = {
  name: string;
  kind: CategoryKind;
};

export type CategoryUpdate = {
  name?: string;
};

export function listCategories(accessToken: string) {
  return apiFetch<Category[]>(
    "/api/v1/categories",
    { method: "GET" },
    { accessToken },
  );
}

export function createCategory(accessToken: string, body: CategoryCreate) {
  return apiFetch<Category>(
    "/api/v1/categories",
    { method: "POST", body: JSON.stringify(body) },
    { accessToken },
  );
}

export function updateCategory(
  accessToken: string,
  id: string,
  body: CategoryUpdate,
) {
  return apiFetch<Category>(
    `/api/v1/categories/${id}`,
    { method: "PATCH", body: JSON.stringify(body) },
    { accessToken },
  );
}

export function deleteCategory(accessToken: string, id: string) {
  return apiFetch<void>(
    `/api/v1/categories/${id}`,
    { method: "DELETE" },
    { accessToken },
  );
}
